source/opt/dead_branch_elim_pass.cpp

```diff
@@ -0,0 +1,271 @@
+      *condVal = false;
+      *condIsConst = true;
+    } break;
+    case SpvOpConstantTrue: {
+      *condVal = true;
+      *condIsConst = true;
+    } break;
+    case SpvOpLogicalNot: {
+      bool negVal;
+      (void)GetConstCondition(cInst->GetSingleWordInOperand(0),
+          &negVal, condIsConst);
+      if (*condIsConst)
+        *condVal = !negVal;
+    } break;
+    default: {
+      *condIsConst = false;
+    } break;
+  }
+}
+
+void DeadBranchElimPass::AddBranch(uint32_t labelId, ir::BasicBlock* bp) {
+  std::unique_ptr<ir::Instruction> newBranch(
+    new ir::Instruction(SpvOpBranch, 0, 0,
+        {{spv_operand_type_t::SPV_OPERAND_TYPE_ID, {labelId}}}));
+  def_use_mgr_->AnalyzeInstDefUse(&*newBranch);
+  bp->AddInstruction(std::move(newBranch));
+}
+
+void DeadBranchElimPass::KillAllInsts(ir::BasicBlock* bp) {
+  bp->ForEachInst([this](ir::Instruction* ip) {
+    def_use_mgr_->KillInst(ip);
+  });
+}
+
+bool DeadBranchElimPass::GetConstConditionalBranch(ir::BasicBlock* bp,
+    ir::Instruction** branchInst, ir::Instruction** mergeInst,
+    bool *condVal) {
+  auto ii = bp->end();
+  --ii;
+  *branchInst = &*ii;
+  if ((*branchInst)->opcode() != SpvOpBranchConditional)
+    return false;
+  --ii;
+  *mergeInst = &*ii;
+  if ((*mergeInst)->opcode() != SpvOpSelectionMerge)
+    return false;
+  bool condIsConst;
+  (void) GetConstCondition(
+      (*branchInst)->GetSingleWordInOperand(kBranchCondConditionalIdInIdx),
+      condVal, &condIsConst);
+  return condIsConst;
+}
+
+bool DeadBranchElimPass::EliminateDeadBranches(ir::Function* func) {
+  // Traverse blocks in structured order
+  std::list<ir::BasicBlock*> structuredOrder;
+  ComputeStructuredOrder(func, &structuredOrder);
+  std::unordered_set<ir::BasicBlock*> elimBlocks;
+  bool modified = false;
+  for (auto bi = structuredOrder.begin(); bi != structuredOrder.end(); ++bi) {
+    // Skip blocks that are already in the elimination set
+    if (elimBlocks.find(*bi) != elimBlocks.end())
+      continue;
+    // Skip blocks that don't have constant conditional branch
+    ir::Instruction* br;
+    ir::Instruction* mergeInst;
+    bool condVal;
+    if (!GetConstConditionalBranch(*bi, &br, &mergeInst, &condVal))
+      continue;
+
+    // Replace conditional branch with unconditional branch
+    const uint32_t trueLabId =
+        br->GetSingleWordInOperand(kBranchCondTrueLabIdInIdx);
+    const uint32_t falseLabId =
+        br->GetSingleWordInOperand(kBranchCondFalseLabIdInIdx);
+    const uint32_t mergeLabId =
+        mergeInst->GetSingleWordInOperand(kSelectionMergeMergeBlockIdInIdx);
+    const uint32_t liveLabId = condVal == true ? trueLabId : falseLabId;
+    const uint32_t deadLabId = condVal == true ? falseLabId : trueLabId;
+    AddBranch(liveLabId, *bi);
+    def_use_mgr_->KillInst(br);
+    def_use_mgr_->KillInst(mergeInst);
+
+    // Iterate to merge block deleting dead blocks
+    std::unordered_set<uint32_t> deadLabIds;
+    deadLabIds.insert(deadLabId);
```

Comment:
Looking at this more, I believe even the code I have written is still not valid. The reason is that the %continuetarget block is still part of the %body select construct and this is invalid because it is not legal for the backedge to come out of a select construct. It is part of the select construct because it is dominated by the %body block and not dominated by the %selectionmerge block.

The big problem here is the OpKill in the %selectionmerge block, which causes the %selectionmerge block to NOT dominate ANY blocks in the loop.

I have tried to come up with a valid version of this code that causes a problem but have not been able to. Please let me know if you think you have one or disagree with my assessment on the validity of this code.